pkg/odo/cli/url/create.go

```diff
@@ -103,167 +76,62 @@ func NewURLCreateOptions() *URLCreateOptions {
-			portList = append(portList, port)
-		}
-		if o.isDocker || o.urlPort == -1 {
-			o.componentPort, err = url.GetValidPortNumber(componentName, o.urlPort, portList)
-			if err != nil {
-				return err
-			}
-		} else {
-			o.componentPort = o.urlPort
-		}
+	isRouteSupported, err := o.Client.IsRouteSupported()
+	if err != nil {
+		return err
+	}
 
-		if len(args) != 0 {
-			o.urlName = args[0]
-		} else {
-			o.urlName = url.GetURLName(componentName, o.componentPort)
-		}
+	var urlType localConfigProvider.URLKind
+	if o.wantIngress || (!isRouteSupported) {
+		urlType = localConfigProvider.INGRESS
+	} else {
+		urlType = localConfigProvider.ROUTE
+	}
 
-		if o.isDocker {
-			o.exposedPort, err = url.GetValidExposedPortNumber(o.exposedPort)
-			if err != nil {
-				return err
-			}
-			o.urlType = envinfo.DOCKER
-		} else {
-			foundContainer := false
-			containerComponents := devObj.Data.GetDevfileContainerComponents()
-			// map TargetPort with containerName
-			containerPortMap := make(map[int]string)
-			for _, component := range containerComponents {
-				if len(o.container) > 0 && !foundContainer {
-					if component.Name == o.container {
-						foundContainer = true
-					}
-				}
-				for _, endpoint := range component.Container.Endpoints {
-					if endpoint.Name == o.urlName {
-						return fmt.Errorf("url %v already exist in devfile endpoint entry under container %v", o.urlName, component.Name)
-					}
-					containerPortMap[int(endpoint.TargetPort)] = component.Name
-				}
-			}
+	// get the name
+	if len(args) != 0 {
+		o.urlName = args[0]
+	}
 
-			if len(o.container) > 0 && !foundContainer {
-				return fmt.Errorf("the container specified: %v does not exist in devfile", o.container)
-			}
-			if containerName, exist := containerPortMap[o.componentPort]; exist {
-				if len(o.container) > 0 && o.container != containerName {
-					return fmt.Errorf("cannot set URL %v under container %v, TargetPort %v is being used under container %v", o.urlName, o.container, o.componentPort, containerName)
-				}
-				o.container = containerName
-			}
-			// container is not provided, or the specified port is not being used under any containers
-			// pick the first container to store the new enpoint
-			if len(o.container) == 0 {
-				o.container = containerComponents[0].Name
-			}
+	// create the localURL
+	o.url = localConfigProvider.LocalURL{
+		Name:      o.urlName,
+		Port:      o.urlPort,
+		Secure:    o.secureURL,
+		Host:      o.host,
+		TLSSecret: o.tlsSecret,
+		Kind:      urlType,
+		Container: o.container,
+		Protocol:  o.protocol,
+		Path:      o.path,
+	}
 
-		}
+	// complete the URL
+	err = o.Context.LocalConfigProvider.CompleteURL(&o.url)
+	if err != nil {
+		return err
+	}
 
-	} else {
-		err = o.InitConfigFromContext()
-		if err != nil {
-			return err
-		}
-		o.componentPort, err = url.GetValidPortNumber(o.Component(), o.urlPort, o.LocalConfigInfo.GetPorts())
+	if o.now {
+		prjName := o.Context.LocalConfigProvider.GetNamespace()
```

Comment:
The underlying function here might return an empty string if the project is not set. That's what I see when I step through the code. But I also understand that an empty value for project/namespace is probably not possible as far as one has kubecfonfig on their system.

So, do you think we should add a check to ensure that `prjName` here is not an empty string? Because if in worst case scenario it is an empty string, we might hit issues later, no?

Reply:
Already handled in `ResolveProject()` later